### PR TITLE
Fixes #245

### LINKED ISSUE #245: Parsing SQLAlchemy URL chokes on `#` or `?` in URL
`reahl.component.dbutils.DatabaseControl` attempts to parse a connection URL on [line 227](https://github.com/reahl/reahl/blob/master/reahl-component/reahl/component/dbutils.py#L227)

However, the call to `urllib_parse.urlparse(url)` barfs on URLs with `#` or `?` in URL. I have updated [the related Python bug](https://bugs.python.org/issue33342) with a link to [some code that demonstrates the problem](https://gist.github.com/metaperl/fc6f43bf6b9a9f874b8f27e29695e68c). 

Out of curiosity why is the database URL parsed instead of simply being passed to SQLAlchemy?

### PR DESCRIPTION
**Pull request by craig-reahl**
Passwords containing special characters (i.e. reserved, punctuation etc.) should be decoded when parsed from the URI

### CODE REVIEW

File: reahl-component/reahl/component_dev/test_dbutils.py [hunk 1]
```diff
@@ -57,20 +59,88 @@ def test_finding_database_control(dbcontrol_fixture):
...
+        self.uri = 'myprefix:///:memory:'
+        self.database_name = ':memory:'
+        self.user_name = None
+        self.password = None
+        self.host = None
+        self.port = None
+
+
+@with_fixtures(DBControlFixture, URIScenarios)
+def test_database_control_settings(dbcontrol_fixture, uri_fixture):
```
Comment (IwanVosloo) [inline reahl-component/reahl/component_dev/test_dbutils.py]: Just for consistency....perhaps call uri_fixture uri_scenario - that's what we do elsewhere.

File: reahl-component/reahl/component_dev/test_dbutils.py [hunk 2]
```diff
@@ -57,20 +59,88 @@ def test_finding_database_control(dbcontrol_fixture):
...
+    assert system_control.db_control.database_name == uri_fixture.database_name
+    assert system_control.db_control.user_name == uri_fixture.user_name
+    assert system_control.db_control.password == uri_fixture.password
+    assert system_control.db_control.host == uri_fixture.host
+    assert system_control.db_control.port == uri_fixture.port
+
+
+@with_fixtures(DBControlFixture)
+def test_invalid_database_control_settings(dbcontrol_fixture):
+    """The minmum an URI should contain
```
Comment (IwanVosloo) [inline reahl-component/reahl/component_dev/test_dbutils.py]: """The minimum an uri should contain is a database name"""

File: reahl-component/reahl/component_dev/test_dbutils.py [hunk 3]
```diff
@@ -57,20 +59,88 @@ def test_finding_database_control(dbcontrol_fixture):
...
+
+    @scenario
+    def complicated(self):
+        self.expected_password = 'p#=sword'
+        self.password = urllib_parse.quote(self.expected_password)
+
+    @scenario
+    def empty(self):
+        self.password = ''
+        self.expected_password = None
```
Comment (craig-reahl) [inline reahl-component/reahl/component_dev/test_dbutils.py]: this tested not needed - is covered in test_database_control_settings

File: reahl-component/reahl/component_dev/test_dbutils.py [hunk 4]
```diff
@@ -57,20 +59,88 @@ def test_finding_database_control(dbcontrol_fixture):
...
+    fixture.config.reahlsystem.connection_uri = 'myprefix:///'
+    with expected(ProgrammerError, test='Please specify a database name in reahlsystem.connection_uri'):
+        SystemControl(fixture.config)
+
+
+class PasswordScenarios(Fixture):
+    @scenario
+    def simple(self):
+        self.password = 'password'
+        self.expected_password = 'password'
```
Comment (craig-reahl) [inline reahl-component/reahl/component_dev/test_dbutils.py]: not needed, covered in test_database_control_settings

File: reahl-component/reahl/component_dev/test_dbutils.py [hunk 5]
```diff
@@ -57,20 +59,88 @@ def test_finding_database_control(dbcontrol_fixture):
...
+        self.password = urllib_parse.quote(self.expected_password)
+
+    @scenario
+    def empty(self):
+        self.password = ''
+        self.expected_password = None
+
+
+@with_fixtures(DBControlFixture, PasswordScenarios)
+def test_database_control_complicated_passwords(dbcontrol_fixture, password_fixture):
```
Comment (craig-reahl) [inline reahl-component/reahl/component_dev/test_dbutils.py]: Remove scenarios, only test complicated scenario, with all parts encoded.

Rename fixture names to scenarios where applicable.

**Pull request merged**